Modified Car Simulation to Use Shared Library

liangfok opened the pull request:
This is in anticipation of adding more car demos, like those that also integrate ROS.

I also made some small grammar fixes and updated the car demo's README to be in anticipating of the future ROS-based car demo.

This PR is a subset of https://github.com/RobotLocomotion/drake/pull/2253, which I'm trying to break down into smaller pieces.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2379) &emsp; Multiple assignees:&emsp;<img alt="@david-german-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17437069?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/david-german-tri">david-german-tri</a>,&emsp;<img alt="@jwnimmer-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596505?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/jwnimmer-tri">jwnimmer-tri</a>
<!-- Reviewable:end -->

liangfok commented:
+@jwnimmer-tri for feature review please
+@david-german-tri for platform review please

Feel free to bounce back or reassign if my assignments are inappropriate.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> #### Modified Car Simulation to Use Shared Library
> 
> This is in anticipation of adding more car demos, like those that also integrate ROS.
> 
> 
> 
> I also make some small grammar fixes and updated the car demo's README to be in anticipating of the future ROS-based car demo.
> 
> 
> 
> This PR is a subset of https://github.com/RobotLocomotion/drake/pull/2253, which I'm trying to break down into smaller pieces.
> 
> 
> </details>


---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KHzQ4v7xZqauQeIfIkl:-1259135408)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Oops. I now notice that this PR includes the changes in `stata_garage_p1.sdf`, which is actually part of https://github.com/RobotLocomotion/drake/pull/2378.  Strange. Looks like I need to polish my git branching skills.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@jwnimmer-tri for feature review please
> +@david-german-tri for platform review please
> 
> Feel free to bounce back or reassign if my assignments are inappropriate.</details>


---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KHzQBv6Ec66vbeJ93Fj:247753756)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed [CI Build 1513](https://drake-jenkins.csail.mit.edu/job/experimental/1513/) and is thus ready for review.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Oops. I now notice that this PR includes the changes in `stata_garage_p1.sdf`, which is actually part of https://github.com/RobotLocomotion/drake/pull/2378.  Strange. Looks like I need to polish my git branching skills.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI0w8FrQkPYut7YwkEJ:1799273877)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 7 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[drake/examples/Cars/car_simulation.cc, line 3 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3ucUFIEDlNDm7BGxR:-KI3ucUFIEDlNDm7BGxS:-2083616953) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L3)):*
> ```cpp
> #include "drake/examples/Cars/car_simulation.h"
> 
> #include <stdlib.h>
> ```

Must be `<cstdlib>` not `<stdlib.h>`.

---

*[drake/examples/Cars/car_simulation.cc, line 11 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qcVJsTG9nvfydAzi:-KI3qcVJsTG9nvfydAzj:160388209) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L11)):*
> ```cpp
> namespace drake {
> 
> std::shared_ptr<RigidBodySystem> CreateRigidBodySystem(int argc,
> ```

TBD

---

*[drake/examples/Cars/car_simulation.cc, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3unK1pNK4Dd_qjtaO:-KI3unK1pNK4Dd_qjtaP:1369251525) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L16)):*
> ```cpp
>     std::cerr << "Usage: " << argv[0] << " vehicle_model [world sdf files ...]"
>               << std::endl;
>     exit(EXIT_FAILURE);
> ```

Functions that call `::exit` are generally a poor design because they remove the ability of calling code to decide how to treat errors.

In this case it's probably okay to leave it this way, but at a minimum the API documentation must document that it calls `::exit`.

---

*[drake/examples/Cars/car_simulation.cc, line 53 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3vyA324z9sHs2U6-9:-KI3vyA324z9sHs2U6-A:-1661220026) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L53)):*
> ```cpp
> 
>   // Instantiates a rigid body system and adds the robot to it.
>   auto rigid_body_sys = std::make_shared<RigidBodySystem>();
> ```

I think this needs allocate_shared, because RBS is Eigen-alignment-brittle.

---

*[drake/examples/Cars/car_simulation.cc, line 60 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3w7MABjrPiY_h4jhQ:-KI3w7MABjrPiY_h4jhR:-15907091) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L60)):*
> ```cpp
>   auto const& tree = rigid_body_sys->getRigidBodyTree();
>   for (int i = 2; i < argc; i++) {
>     if (std::string(argv[i]) == "--duration") {
> ```

This is too brittle / magic / undocumented in the API.

I wonder if the easiest path forward is to just pass a `vector<string>` to this method, and let the main program sort out which options go where?

---

*[drake/examples/Cars/car_simulation.cc, line 67 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3vf24vmDC90zWAn7m:-KI3vf24vmDC90zWAn7n:317146203) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L67)):*
> ```cpp
>   }
> 
>   // If no environment is specified, the following code adds a flat terrain.
> ```

This is missing from the API docs.

---

*[drake/examples/Cars/car_simulation.cc, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3sWWER3nI1kJlbM-t:-KI3sWWER3nI1kJlbM-u:-1685313395) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L99)):*
> ```cpp
>     PDControlSystem<RigidBodySystem>>>
> CreateVehicleSystem(std::shared_ptr<RigidBodySystem> rigid_body_sys) {
>   auto const& tree = rigid_body_sys->getRigidBodyTree();
> ```

'const auto&' per cppguide

---

*[drake/examples/Cars/car_simulation.cc, line 108 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3tNbOEVslQO8IWqyx:-KI3tNbOEVslQO8IWqyy:340308258) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L108)):*
> ```cpp
> 
>   // Sets up PD controllers for throttle and steering.
>   double kpSteering = 400, kdSteering = 80, kThrottle = 100;
> ```

Missing 'const'.  (Ordinarily this is not a defect, but for kFoo constants it is.)

Or possibly the defect here is that the variable is actually kp_steering (where k is coming from a well-known math variable in the problem domain, not a Konstant.)

---

*[drake/examples/Cars/car_simulation.cc, line 118 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3u2nUN4CCaAPNvkdE:-KI3u2nUN4CCaAPNvkdF:-1685313395) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L118)):*
> ```cpp
> 
>     if (actuator_name == "steering") {
>       auto const& b = tree->actuators[actuator_idx].body;
> ```

'const auto&' per cppguide

---

*[drake/examples/Cars/car_simulation.cc, line 121 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3uFS3JF9wPpYz84-S:-KI3uFS4hjTGuj4iiolj:-1128665501) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L121)):*
> ```cpp
>       Kp(actuator_idx, b->position_num_start) = kpSteering;  // steering
>       Kd(actuator_idx, b->velocity_num_start) = kdSteering;  // steeringdot
>       map_driving_cmd_to_x_d(b->position_num_start, 0) = 1;  // steering command
> ```

I think the `0` here is a brittle magic number; I think its really `DrivingCommandIndices::kSteeringAngle`.

---

*[drake/examples/Cars/car_simulation.cc, line 125 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3u742Rhp-Gm1nkQoW:-KI3u742Rhp-Gm1nkQoX:-1685313395) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L125)):*
> ```cpp
>     } else if (actuator_name == "right_wheel_joint" ||
>                actuator_name == "left_wheel_joint") {
>       auto const& b = tree->actuators[actuator_idx].body;
> ```

'const auto&' per cppguide

---

*[drake/examples/Cars/car_simulation.cc, line 129 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3uRDfe4jVLVnk3Dsj:-KI3uRDfe4jVLVnk3Dsk:-97629083) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L129)):*
> ```cpp
>       map_driving_cmd_to_x_d(
>           tree->number_of_positions() + b->velocity_num_start, 1) =
>           20;  // throttle (velocity) command
> ```

I think the `0` here is a brittle magic number; I think its really `DrivingCommandIndices::kThrottle`.

---

*[drake/examples/Cars/car_simulation.cc, line 132 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3uRj2K_7c32hBFg32:-KI3uRj2K_7c32hBFg33:-1336201364) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L132)):*
> ```cpp
>       map_driving_cmd_to_x_d(
>           tree->number_of_positions() + b->velocity_num_start, 2) =
>           -20;  // braking (velocity) command
> ```

I think the `0` here is a brittle magic number; I think its really `DrivingCommandIndices::kBrake`.

---

*[drake/examples/Cars/car_simulation.cc, line 140 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3vLK47P-WZg8VMDbW:-KI3vLK47P-WZg8VMDbX:1252015810) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L140)):*
> ```cpp
> 
>   auto vehicle_sys = cascade(
>       std::make_shared<
> ```

I think this needs allocated_shared, because Gain is Eigen-alignment-brittle.

---

*[drake/examples/Cars/car_simulation.cc, line 152 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3s9VVI-nNWbpDwg1s:-KI3s9VVI-nNWbpDwg1t:279976922) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L152)):*
> ```cpp
>   if (sim_options == nullptr)
>     throw std::runtime_error("ERROR: drake::SetSimulationOptions(): "
>       "Simulation options are null!");
> ```

Multi-line bodies require braces per cppguide.

---

*[drake/examples/Cars/car_simulation.cc, line 159 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qwKtYmklE8oYgMNC:-KI3qwKwkBZATam5wAeu:-1685313395) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L159)):*
> ```cpp
> 
> VectorXd GetInitialState(std::shared_ptr<RigidBodySystem> rigid_body_sys) {
>   auto const& tree = rigid_body_sys->getRigidBodyTree();
> ```

'const auto&' per cppguide

---

*[drake/examples/Cars/car_simulation.h, line 65 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3mfJxbxrXPaHU-QKS:-KI3mfJxbxrXPaHU-QKT:-1353631379) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.h#L65)):*
> ```objc
> void SetSimulationOptions(SimulationOptions* sim_options,
>   double initial_step_size = 5e-3,
>   double timeout_seconds = std::numeric_limits<double>::infinity());
> ```

I'll suggest that this API is complicated and easy to mis-use.  The possibly-null pointer is a headache, and the positional arguments to alter the defaults are going to be hard to read at call sites.

I can offer this as one possible replacement:
.h: `SimulationOptions GetCarSimulationDefaultOptions();`
.cc: `{
  SimulationOptions result = Drake::default_simulation_options;
  result.initial_step_size = 5e-3;
  result.timeout_seconds = ...;
  return result;
}`
call site: `
  SimulationOptions options = GetCarSimulationDefaultOptions();
`

If calling code wants to use values other than the defaults, it can assign to options using its named fields.

---

*[drake/examples/Cars/car_simulation.h, line 75 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qMzWmiY5YHiRnAYC:-KI3qMzWmiY5YHiRnAYD:-673687637) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.h#L75)):*
> ```objc
> DRAKECARSIMLIB_EXPORT
> Eigen::VectorXd GetInitialState(
>     std::shared_ptr<RigidBodySystem> rigid_body_sys);
> ```

No need to pass by shared_ptr; suggest pass-by-const-reference.

---

*[drake/examples/Cars/CMakeLists.txt, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI-8dfHI-7Gr3NBKYHH:-KI-8dfHI-7Gr3NBKYHI:-239822330) ([raw file](https://github.com/robotlocomotion/drake/blob/50e7ac4c336a409d3b0e5ab1973e6d97fbe439e6/drake/examples/Cars/CMakeLists.txt#L22)):*
> ```CMake
>   target_link_libraries(simple_car_demo drakeCars drakeRBSystem drakeLCMSystem)
> 
>   add_library_with_exports(LIB_NAME drakeCarSimLib SOURCE_FILES car_simulation.cc)
> ```

A whole new library seems at first glance like excessive baggage for calling code to worry about, and developers to maintain.  Why not put this file into the drakeCars library?  My vision for that library is to contain all of the non-main code from the Cars directory.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I've addressed or at least responded to all reviewer comments. PTAL.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR passed [CI Build 1513](https://drake-jenkins.csail.mit.edu/job/experimental/1513/) and is thus ready for review.
</details>


---

Review status: 1 of 6 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 3 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3ucUFIEDlNDm7BGxR:-KI42dc96Hs9XEtw52uC:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L3)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Must be `<cstdlib>` not `<stdlib.h>`.
</details>
Done.

---

*[drake/examples/Cars/car_simulation.cc, line 11 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qcVJsTG9nvfydAzi:-KI462HZiR-5Cy6BfsC0:32991569) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L11)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> TBD
</details>
To Be Determined?

---

*[drake/examples/Cars/car_simulation.cc, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3unK1pNK4Dd_qjtaO:-KI49BKtkVbGyM__ofio:-1940796319) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L16)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Functions that call `::exit` are generally a poor design because they remove the ability of calling code to decide how to treat errors.
> 
> In this case it's probably okay to leave it this way, but at a minimum the API documentation must document that it calls `::exit`.
</details>
I added documentation to the method declaration in `car_simulation.h` indicating that this method calls `::exit(EXIT_FAILURE)`. 

Unfortunately, I wasn't able to verify that Doxygen renders the description correctly because of https://github.com/RobotLocomotion/drake/issues/2393. 

---

*[drake/examples/Cars/car_simulation.cc, line 53 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3vyA324z9sHs2U6-9:-KI4AG3Jo57Av2AtqORm:-795674665) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L53)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think this needs allocate_shared, because RBS is Eigen-alignment-brittle.
</details>
Fixed. Thanks!

---

*[drake/examples/Cars/car_simulation.cc, line 60 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3w7MABjrPiY_h4jhQ:-KI4AZtXgtEWGxAn3SC5:178775331) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L60)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This is too brittle / magic / undocumented in the API.
> 
> I wonder if the easiest path forward is to just pass a `vector<string>` to this method, and let the main program sort out which options go where?
</details>
We should probably eventually switch to use [GFlags](https://github.com/gflags/gflags).
Would making that switch in this PR result in this PR being too big?

I modified the print statement describing the command line options to include the `--duration [duration in seconds]` flag.  

---

*[drake/examples/Cars/car_simulation.cc, line 67 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3vf24vmDC90zWAn7m:-KI4C_2XjdSpPr-QM1DU:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L67)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This is missing from the API docs.
</details>
Done.

---

*[drake/examples/Cars/car_simulation.cc, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3sWWER3nI1kJlbM-t:-KI4Ci68lnjH3irTuAum:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> 'const auto&' per cppguide
</details>
Fixed. Thanks.

---

*[drake/examples/Cars/car_simulation.cc, line 108 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3tNbOEVslQO8IWqyx:-KI4CnXf4jsUu7f-dL66:2136349459) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L108)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing 'const'.  (Ordinarily this is not a defect, but for kFoo constants it is.)
> 
> Or possibly the defect here is that the variable is actually kp_steering (where k is coming from a well-known math variable in the problem domain, not a Konstant.)
</details>
Fixed. Thanks. Eventually I'd like to make these input parameters specified by the user.

---

*[drake/examples/Cars/car_simulation.cc, line 118 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3u2nUN4CCaAPNvkdE:-KI4CusvBBvNWFdAiLXn:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L118)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> 'const auto&' per cppguide
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 121 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3uFS3JF9wPpYz84-S:-KI4LStnpKHuwor2vbTE:1358578923) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L121)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think the `0` here is a brittle magic number; I think its really `DrivingCommandIndices::kSteeringAngle`.
</details>
I believe you're right. Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 125 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3u742Rhp-Gm1nkQoW:-KI4KourTvrugEvTLLK5:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L125)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> 'const auto&' per cppguide
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 129 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3uRDfe4jVLVnk3Dsj:-KI4N1-lO-o3ArdfDRNP:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think the `0` here is a brittle magic number; I think its really `DrivingCommandIndices::kThrottle`.
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 132 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3uRj2K_7c32hBFg32:-KI4N00LpADHyjhw97qc:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L132)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think the `0` here is a brittle magic number; I think its really `DrivingCommandIndices::kBrake`.
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 140 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3vLK47P-WZg8VMDbW:-KI4NeCMsqHqjct34wIH:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L140)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think this needs allocated_shared, because Gain is Eigen-alignment-brittle.
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 152 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3s9VVI-nNWbpDwg1s:-KI4NjtlNyXyfhLusktT:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L152)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Multi-line bodies require braces per cppguide.
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.cc, line 159 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qwKtYmklE8oYgMNC:-KI4NmudGVrIqFjoOsP0:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L159)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> 'const auto&' per cppguide
</details>
Fixed.

---

*[drake/examples/Cars/car_simulation.h, line 65 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3mfJxbxrXPaHU-QKS:-KI4TxDHA64ybT8usyaZ:-755327526) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.h#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I'll suggest that this API is complicated and easy to mis-use.  The possibly-null pointer is a headache, and the positional arguments to alter the defaults are going to be hard to read at call sites.
> 
> I can offer this as one possible replacement:
> .h: `SimulationOptions GetCarSimulationDefaultOptions();`
> .cc: `{
>   SimulationOptions result = Drake::default_simulation_options;
>   result.initial_step_size = 5e-3;
>   result.timeout_seconds = ...;
>   return result;
> }`
> call site: `
>   SimulationOptions options = GetCarSimulationDefaultOptions();
> `
> 
> If calling code wants to use values other than the defaults, it can assign to options using its named fields.
> 
</details>
Agreed. Suggested changes applied.

---

*[drake/examples/Cars/car_simulation.h, line 75 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qMzWmiY5YHiRnAYC:-KI4PYZXH1QyrjoG0Ol7:-474899032) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.h#L75)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> No need to pass by shared_ptr; suggest pass-by-const-reference.
</details>
Changed as suggested.

---

*[drake/examples/Cars/CMakeLists.txt, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI-8dfHI-7Gr3NBKYHH:-KI4Pi3ZdZOzHvGhQ-Z7:286237144) ([raw file](https://github.com/robotlocomotion/drake/blob/50e7ac4c336a409d3b0e5ab1973e6d97fbe439e6/drake/examples/Cars/CMakeLists.txt#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> A whole new library seems at first glance like excessive baggage for calling code to worry about, and developers to maintain.  Why not put this file into the drakeCars library?  My vision for that library is to contain all of the non-main code from the Cars directory.
</details>
When I originally wrote this code, I believe `drakeCars` was called `drakeSimpleCars` or something like that, which is why I created a separate library. Fixed in the next revision by merging the `drakeCarSimLib` library into the `drakeCars`library.

Style question: Wouldn't it be more informative if the library names ended with the word "Library" or "Lib" in it, as in `drakeCarsLib`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI4UvJmhVRobcUdTa7R:1086381844)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Cars/car_sim_lcm.cpp, line 43 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI4tD9EX30yOQbYsRiA:-KI4tD9EX30yOQbYsRiB:1634697953) ([raw file](https://github.com/robotlocomotion/drake/blob/d5ee71b6d4ba59248fb0bc48769f497a054c8fd5/drake/examples/Cars/car_sim_lcm.cpp#L43)):*
> ```cpp
>   for (int i = 2; i < argc; i++) {
>     if (std::string(argv[i]) == "--duration") {
>       duration = atof(argv[++i]);
> ```

I think if `--duration` is the final command line argument (user error), then this would read past the end of the `argv` array.

---

*[drake/examples/Cars/car_simulation.cc, line 11 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3qcVJsTG9nvfydAzi:-KI4qNfOXEOzzB5XyZ3J:355821602) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> To Be Determined?
</details>
Sorry, I thought I removed all of those.  No action required.  (It's a note I use for myself to mark which functions I've reviewed yet or not.)

---

*[drake/examples/Cars/car_simulation.cc, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3unK1pNK4Dd_qjtaO:-KI4qXrCPCZ-mYgWibl5:543513016) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L16)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I added documentation to the method declaration in `car_simulation.h` indicating that this method calls `::exit(EXIT_FAILURE)`. 
> 
> Unfortunately, I wasn't able to verify that Doxygen renders the description correctly because of https://github.com/RobotLocomotion/drake/issues/2393. 
</details>
Fine by me.  I actually think that developers testing doxygen is a waste of time and never do it.

---

*[drake/examples/Cars/car_simulation.cc, line 60 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3w7MABjrPiY_h4jhQ:-KI4qtlQwk8bEqAZgbcG:-684977213) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L60)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> We should probably eventually switch to use [GFlags](https://github.com/gflags/gflags).
> Would making that switch in this PR result in this PR being too big?
> 
> I modified the print statement describing the command line options to include the `--duration [duration in seconds]` flag.  
</details>
No, I don't think gflags in this PR is worthwhile (though per f2f one of your or I might do it soon in another PR).

So I think my objection in this case will be to brittleness / maintainability.  That the magic one-special command-line option is named both here an in the car_sim_lcm.cc file.

Possible resolutions:
- In both places that "--duration" is mentioned, add a TODO comment that _this string is duplicated in [other filename] and be sure to update both places at once_.  (This option is probably the best choice.)
- Factor out the "duration parsing" out of car_sim_lcm and into car_simulation.h.
- Change car_sim_lcm to pass only a portion of argv to the rigid body loading method.

---

*[drake/examples/Cars/CMakeLists.txt, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI-8dfHI-7Gr3NBKYHH:-KI4rzSTHFnEx0mVEzmT:1270162242) ([raw file](https://github.com/robotlocomotion/drake/blob/50e7ac4c336a409d3b0e5ab1973e6d97fbe439e6/drake/examples/Cars/CMakeLists.txt#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> When I originally wrote this code, I believe `drakeCars` was called `drakeSimpleCars` or something like that, which is why I created a separate library. Fixed in the next revision by merging the `drakeCarSimLib` library into the `drakeCars`library.
> 
> Style question: Wouldn't it be more informative if the library names ended with the word "Library" or "Lib" in it, as in `drakeCarsLib`?
</details>
Yup, the library name used to be overly specific, so certainly would have been weird to put the new code into a simple car library.

I think in most places, the lack of "lib" is fairly clear -- files end in .dll or .so, link flags are -ldrakeCars, etc.  Just in CMakeLists here is hard to understand, where what each noun means is not always obvious, but IMO that's just life with "variables and filenames must be identical" from CMake land, and not worth making "libdrakeCarsLib.so" appear, where we would become a source of ridicule for our downstream users.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 3 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_sim_lcm.cpp, line 43 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI4tD9EX30yOQbYsRiA:-KI52AAmg03nd5tnzobo:-1221891284) ([raw file](https://github.com/robotlocomotion/drake/blob/d5ee71b6d4ba59248fb0bc48769f497a054c8fd5/drake/examples/Cars/car_sim_lcm.cpp#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think if `--duration` is the final command line argument (user error), then this would read past the end of the `argv` array.
</details>
Right. Fixed in next revision.

---

*[drake/examples/Cars/car_simulation.cc, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3unK1pNK4Dd_qjtaO:-KI52HN0RaB_0X5HqqMW:-660659362) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L16)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Fine by me.  I actually think that developers testing doxygen is a waste of time and never do it.
</details>
Should that be handled by the "customer relations" department once we have such a thing?

Really curious what @sherm1 has to say :smile: 

---

*[drake/examples/Cars/car_simulation.cc, line 60 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3w7MABjrPiY_h4jhQ:-KI53maS2Gt8QkKKHFm8:1201734310) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L60)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> No, I don't think gflags in this PR is worthwhile (though per f2f one of your or I might do it soon in another PR).
> 
> So I think my objection in this case will be to brittleness / maintainability.  That the magic one-special command-line option is named both here an in the car_sim_lcm.cc file.
> 
> Possible resolutions:
> - In both places that "--duration" is mentioned, add a TODO comment that _this string is duplicated in [other filename] and be sure to update both places at once_.  (This option is probably the best choice.)
> - Factor out the "duration parsing" out of car_sim_lcm and into car_simulation.h.
> - Change car_sim_lcm to pass only a portion of argv to the rigid body loading method.
</details>
I refactored the code by moving the `--duration` parsing logic into `car_simulation.h/.cc`.

---

*[drake/examples/Cars/CMakeLists.txt, line 22 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI-8dfHI-7Gr3NBKYHH:-KI53yr51qUhjgaXhhDu:-2037353401) ([raw file](https://github.com/robotlocomotion/drake/blob/50e7ac4c336a409d3b0e5ab1973e6d97fbe439e6/drake/examples/Cars/CMakeLists.txt#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Yup, the library name used to be overly specific, so certainly would have been weird to put the new code into a simple car library.
> 
> I think in most places, the lack of "lib" is fairly clear -- files end in .dll or .so, link flags are -ldrakeCars, etc.  Just in CMakeLists here is hard to understand, where what each noun means is not always obvious, but IMO that's just life with "variables and filenames must be identical" from CMake land, and not worth making "libdrakeCarsLib.so" appear, where we would become a source of ridicule for our downstream users.
</details>
OK no "Lib" postfix for libraries then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Addressed all comments. PTAL.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I've addressed or at least responded to all reviewer comments. PTAL.
</details>


---

Review status: 3 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI56ttzPrM-ccylXVse:2038951528)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 3 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.cc, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3unK1pNK4Dd_qjtaO:-KI5JoxiwCMHCoCWqBMO:2019012500) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L16)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Should that be handled by the "customer relations" department once we have such a thing?
> 
> Really curious what @sherm1 has to say :smile: 
</details>
Ha! Are trying to stir up trouble, Liang? I strongly disagree with Jeremy on this -- doxygen is code (typically buggy) and more importantly it is generating part of the user-visible API. So of course it should be tested! But I wasn't going to inject that discussion into this PR and will now crawl back into my hole.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_sim_lcm.cpp, line 26 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI5LeIX01NNrc88BeBX:-KI5LeIYiQ06gSem3SAE:1388093926) ([raw file](https://github.com/robotlocomotion/drake/blob/802e111943a38c0d6f6dd844738b156478b16828/drake/examples/Cars/car_sim_lcm.cpp#L26)):*
> ```cpp
>   // Instantiates a duration variable that will be set by the call to
>   // drake::CreateRigidBodySystem() below.
>   double duration;
> ```

"initialize variables in the declaration" per cppguide

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Addressed all reviewer comments. PTAL Thanks!

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Addressed all comments. PTAL.
</details>


---

Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/examples/Cars/car_sim_lcm.cpp, line 26 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI5LeIX01NNrc88BeBX:-KI5RPA2HXIFL_1wsBEZ:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/802e111943a38c0d6f6dd844738b156478b16828/drake/examples/Cars/car_sim_lcm.cpp#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> "initialize variables in the declaration" per cppguide
</details>
Fixed. Thanks.

---

*[drake/examples/Cars/car_simulation.cc, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI3unK1pNK4Dd_qjtaO:-KI5RQJiHlEG5swBFTdq:2024536739) ([raw file](https://github.com/robotlocomotion/drake/blob/7cc0a93a1d505e84cfb9357066567fe8a5bab879/drake/examples/Cars/car_simulation.cc#L16)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Ha! Are trying to stir up trouble, Liang? I strongly disagree with Jeremy on this -- doxygen is code (typically buggy) and more importantly it is generating part of the user-visible API. So of course it should be tested! But I wasn't going to inject that discussion into this PR and will now crawl back into my hole.
</details>
Nah. Just trying to stimulate discussion. 

I agree _this_ particular discussion shouldn't be done as margin notes inside a PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI5RfdqRP_ORqpIR25b:-840608521)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Feature <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.  I also marked sherm's comment as "satisfied" so we have a clean slate for the platform reviewer.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Addressed all reviewer comments. PTAL Thanks!
</details>


---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI5S5jIsOG-Xpeg73ka:2015283548)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@liangfok Oops, looks like #2371 probably has a minor merge conflict with this PR; you'll need to resolve that also.

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I resolved the conflicts and am waiting for the CI build servers to report back. I wish the CI would e-mail, text, or slack me when it's done.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> @liangfok Oops, looks like #2371 probably has a minor merge conflict with this PR; you'll need to resolve that also.
</details>


---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI5VnrsekqxQuJa1cXn:-1792686092)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Strange. Both 32-bit and 64-bit Windows indicate a problem with unit test `functional_form_test`. Unfortunately, the console logs don't offer much info:

```
21:20:35 34/41 Test  #3: functional_form_test .....................***Exception: Other  0.61 sec
```

For more info: https://drake-jenkins.csail.mit.edu/job/experimental/1560/

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I resolved the conflicts and am waiting for the CI build servers to report back. I wish the CI would e-mail, text, or slack me when it's done.
</details>


---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI5ofJ2udp5Vhwi7CuR:-2075085891)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Wow.  Somehow PRs are using ninja on windows now.  I knew that was a work in progress in experimental land, but I'm sad that it got turned on without us hearing about the configuration change.  Maybe it was an accident?

I would say ignore it for now; this PR needs platform review anyway before merge.  We can placate jenkins after that is done, if we still need to.

liangfok commented:
This PR is ready for final review / merge.

jwnimmer-tri commented:
Well, I am still LGTM with it, but still needs +@david-german-tri platform review.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR is ready for final review / merge.
</details>


---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9DkSTgpgfoyIaSaG-:-1372683422)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I don't really think my review was necessary in this case, but I did notice one thing.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Well, I am still LGTM with it, but still needs +@david-german-tri platform review.
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[drake/examples/Cars/car_simulation.h, line 22 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI9MbHLZEHca9yxKl5k:-KI9MbHLZEHca9yxKl5l:-1149024022) ([raw file](https://github.com/robotlocomotion/drake/blob/089d7493abf419483d6bb33ff021e5c53bd2fbbc/drake/examples/Cars/car_simulation.h#L22)):*
> ```objc
> using Drake::SimulationOptions;
> 
> namespace drake {
> ```

I think this needs an additional layer of namespacing.  `cars`, perhaps?

---

*[drake/examples/Cars/README.md, line 80 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI9NEZofhpSAwRpuV2c:-KI9NEZofhpSAwRpuV2d:1677689894) ([raw file](https://github.com/robotlocomotion/drake/blob/089d7493abf419483d6bb33ff021e5c53bd2fbbc/drake/examples/Cars/README.md#L80)):*
> ```Markdown
> ### Simulation Using Drake + LCM + ROS
> 
> *Comming Soon!*
> ```

Coming

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9NW8EBp6IGTdUP02Y:-743365961)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I think for any amount of meaningful code (vs doc, nits), two pairs of eyes is a best practice.  But perhaps another feature reviewer would have been good enough.

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> :lgtm: I don't really think my review was necessary in this case, but I did notice one thing.
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9ORmxd7QA-r_e6y7k:-1912018289)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks for the review comments. Working on addressing them now.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think for any amount of meaningful code (vs doc, nits), two pairs of eyes is a best practice.  But perhaps another feature reviewer would have been good enough.
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9Pjow3PeI4f8uVZqU:545274182)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Thanks for the review comments. Working on addressing them now.
</details>


---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.h, line 22 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI9MbHLZEHca9yxKl5k:-KI9UjFS9T_9hAbsc0Gz:-1044564391) ([raw file](https://github.com/robotlocomotion/drake/blob/089d7493abf419483d6bb33ff021e5c53bd2fbbc/drake/examples/Cars/car_simulation.h#L22)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I think this needs an additional layer of namespacing.  `cars`, perhaps?
</details>
Good idea. I added two additional name spaces to keep it consistent with the directory structure and highlight the fact that this is example code.

---

*[drake/examples/Cars/README.md, line 80 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI9NEZofhpSAwRpuV2c:-KI9Q2Y0Pzj4VQQLrNl-:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/089d7493abf419483d6bb33ff021e5c53bd2fbbc/drake/examples/Cars/README.md#L80)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Coming
</details>
Fixed. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9UupUufBY_TugO-Fq:701516011)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> platform

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL. Thanks.
</details>


---

Reviewed 1 of 7 files at r1, 1 of 1 files at r6, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Cars/car_simulation.h, line 22 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI9MbHLZEHca9yxKl5k:-KI9YlffotZ6ku7_2yH2:676834233) ([raw file](https://github.com/robotlocomotion/drake/blob/089d7493abf419483d6bb33ff021e5c53bd2fbbc/drake/examples/Cars/car_simulation.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Good idea. I added two additional name spaces to keep it consistent with the directory structure and highlight the fact that this is example code.
</details>
LGTM

---

*[drake/examples/Cars/README.md, line 80 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-KI9NEZofhpSAwRpuV2c:-KI9YpI-Kvoh4OUuaFzP:676834233) ([raw file](https://github.com/robotlocomotion/drake/blob/089d7493abf419483d6bb33ff021e5c53bd2fbbc/drake/examples/Cars/README.md#L80)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Fixed. Thanks.
</details>
LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9i1eFoLM8UWUU7rCO:769217762)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Just waiting for jenkins now.

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> :lgtm: platform
</details>


---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9inGkPljtmZTPFKH3:805858907)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed [Jenkins build 1582](https://drake-jenkins.csail.mit.edu/job/experimental/1582/) and this ready for final review / merge.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Just waiting for jenkins now.
</details>


---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2379#-:-KI9mbzSAuX10kj8Ti_F:-1934531572)*
<!-- Sent from Reviewable.io -->
